wasm-runtime/polkadot/src/support/environment.rs

```diff
@@ -0,0 +1,33 @@
+use runtime_support::{Rc, RefCell, transmute, Box};
+use primitives::{BlockNumber, Digest};
+
+#[derive(Default)]
+pub struct Environment {
+	pub block_number: BlockNumber,
+	pub digest: Digest,
+	pub next_log_index: usize,
+}
+
+pub fn with_env<T, F: FnOnce(&mut Environment) -> T>(f: F) -> T {
+	let e = env();
+	let mut eb = e.borrow_mut();
+	f(&mut *eb)
+}
+
+pub fn env() -> Rc<RefCell<Environment>> {
+	// Initialize it to a null value
+	static mut SINGLETON: *const Rc<RefCell<Environment>> = 0 as *const Rc<RefCell<Environment>>;
+
+	unsafe {
+		if SINGLETON == 0 as *const Rc<RefCell<Environment>> {
+			// Make it
+			let singleton: Rc<RefCell<Environment>> = Rc::new(RefCell::new(Default::default()));
+
+			// Put it in the heap so it can outlive this call
+			SINGLETON = transmute(Box::new(singleton));
+		}
+
+		// Now we give out a copy of the data that is safe to use concurrently.
```

Comment:
it's really not. Neither `Rc` nor `RefCell` is safe under concurrency and I'm fairly sure that this `env` function can be used to send them to different threads.

Reply:
also there could be a potential race between threads initializing the singleton. Take a look at `std::sync::ONCE`

Reply:
the wasm code is guaranteed to be single-threaded; i don't want to bring in non no-std crates until i know more about the restrictions we'll be under for wasm.

Reply:
then there should be some kind of test or protection which prevents this code from compiling in any other scenario

Reply:
the current code has a thread-safe variant (for when running tests); is there some way to detect the wasm target to ensure the non-thread-safe version only compiles when its wasm?

Reply:
`#[cfg(target_arch = "wasm32")]`